src/plugins/hidsim/gui/vmi_win_gui_parser.cpp

```diff
@@ -0,0 +1,1040 @@
+            bit_idx = x % 8;
+            bit = 0x80 >> bit_idx;
+            map[byte] |= bit;
+        }
+    }
+}
+
+
+/*
+ * Determine, if windows is visible;
+ * important since invisible windows might have visible children
+ */
+bool is_wnd_visible(vmi_instance_t vmi, vmi_pid_t pid, addr_t wnd)
+{
+    uint32_t style = 0;
+
+    if (VMI_FAILURE == vmi_read_32_va(vmi, wnd + symbol_offsets.wnd_style, pid, (uint32_t*)&style))
+        return false;
+
+    if (style & WS_VISIBLE)
+        return true;
+
+    return false;
+}
+
+struct wnd* construct_wnd_container(vmi_instance_t vmi, vmi_pid_t pid, addr_t win)
+{
+    status_t ret = VMI_FAILURE;
+
+    uint32_t x0 = 0;
+    uint32_t x1 = 0;
+    uint32_t y0 = 0;
+    uint32_t y1 = 0;
+
+    uint32_t rx0 = 0;
+    uint32_t rx1 = 0;
+    uint32_t ry0 = 0;
+    uint32_t ry1 = 0;
+
+    uint32_t style = 0;
+    uint32_t exstyle = 0;
+    wchar_t* wn = NULL;
+
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_wnd_offset + symbol_offsets.rc_left_offset, pid, (uint32_t*)&x0);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_wnd_offset + symbol_offsets.rc_right_offset, pid, (uint32_t*)&x1);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_wnd_offset + symbol_offsets.rc_top_offset, pid, (uint32_t*)&y0);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_wnd_offset + symbol_offsets.rc_bottom_offset, pid, (uint32_t*)&y1);
+
+    /* Determine, if windows is visible */
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.wnd_style, pid, (uint32_t*)&style);
+
+    /* Determine extended style attributes */
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.wnd_exstyle, pid, (uint32_t*)&exstyle);
+
+    /* Retrieves atom value */
+    addr_t pcls = 0;
+    ret = vmi_read_addr_va(vmi, win + symbol_offsets.pcls_offset, pid, &pcls);
+    uint16_t atom = 0;
+    ret = vmi_read_16_va(vmi, pcls + symbol_offsets.cls_atom_offset, pid, &atom);
+
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_client_offset + symbol_offsets.rc_left_offset, pid, (uint32_t*)&rx0);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_client_offset + symbol_offsets.rc_right_offset, pid, (uint32_t*)&rx1);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_client_offset + symbol_offsets.rc_top_offset, pid, (uint32_t*)&ry0);
+    ret = vmi_read_32_va(vmi, win + symbol_offsets.rc_client_offset + symbol_offsets.rc_bottom_offset, pid, (uint32_t*)&ry1);
+
+    addr_t str_name_off;
+
+    /* Retrieves window name */
+    if (VMI_FAILURE != vmi_read_addr_va(vmi, win + symbol_offsets.wnd_strname_offset +
+            symbol_offsets.large_unicode_buf_offset, pid, &str_name_off))
+    {
+        /* Length is always 0, therefore always read 255 chars */
+        wn = read_wchar_str_pid(vmi, str_name_off, (size_t)255, pid);
+    }
+
+    if (ret == VMI_FAILURE)
+        return NULL;
+
+    /* Populate struct, if no failure occured */
+    struct wnd* wc = (struct wnd*) calloc(1, sizeof(struct wnd));
```

Comment:
Maybe allocate the structure first, and copy the data directly instead of using temporaries?

Reply:
Changed in 6a7a9c8fde8f434abbd20b8f1ee3608c42468556